SLS-913 Set base and backlink checkpoint LSNs during page write

pmacko86 opened the pull request:
Set base and backlink LSNs for pages. I just went to all the places that use or set `base_checkpoint_id` and `backling_checkpoint_id` and updated those.

ddanderson commented on src/block_disagg/block_disagg_read.c:
```diff
@@ -196,6 +196,8 @@ __block_disagg_read_multiple(WT_SESSION_IMPL *session, WT_BLOCK_DISAGG *block_di
                     block_meta->page_id = page_id;
                     block_meta->checkpoint_id = checkpoint_id;
                     block_meta->reconciliation_id = reconciliation_id;
+                    block_meta->backlink_lsn = get_args.backlink_lsn;
+                    block_meta->base_lsn = get_args.base_lsn;
```
We might also need a fix in PALM, as it doesn't already set `get_args.base_lsn` or `get_args.backlink_lsn`.  Maybe this:
```
diff --git a/ext/page_log/palm/palm.c b/ext/page_log/palm/palm.c
index d2f1a558a..e3ddefe98 100644
--- a/ext/page_log/palm/palm.c
+++ b/ext/page_log/palm/palm.c
@@ -625,6 +625,15 @@ palm_handle_get(WT_PAGE_LOG_HANDLE *plh, WT_SESSION *session, uint64_t page_id,
         memset(&results_array[count], 0, sizeof(WT_ITEM));
         PALM_KV_ERR(palm, session, palm_resize_item(&results_array[count], matches.size));
         memcpy(results_array[count].mem, matches.data, matches.size);
+        if (count > 0) {
+            if (get_args->backlink_checkpoint_id != matches.backlink)
+                PALM_KV_ERR(palm, session, EINVAL);
+            if (get_args->base_checkpoint_id != matches.base)
+                PALM_KV_ERR(palm, session, EINVAL);
+        } else {
+            get_args->backlink_checkpoint_id = matches.backlink;
+            get_args->base_checkpoint_id = matches.base;
+        }
     }
     /* Did the caller give us enough output entries to hold all the results? */
     if (count == *results_count && palm_kv_next_page_match(&matches))
```

pmacko86 replied:
Thanks for pointing this out! I forgot about this part; I was just searching for strings such as `backlink_checkpoint_id`, so I missed it.

Fixing this turned out to be a bigger change. Please see my other comment. Thank you!

quchenhao commented on src/reconcile/rec_write.c:
```diff
@@ -2445,7 +2445,8 @@ __rec_split_write(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_REC_CHUNK *chunk
         if (checkpoint_id != multi->block_meta.checkpoint_id) {
             WT_ASSERT(session, checkpoint_id > multi->block_meta.checkpoint_id);
-            /* Delta reuse the previous base checkpoint id. */
+            /* Delta reuse the previous base checkpoint id and lsn. */
```
It should be base lsn?

pmacko86 commented:
As Don pointed out, my first attempt missed implementing backlink and base LSNs in PALM. In my last commit, I:
- implemented backlink and base LSNs in PALM
- added validation checks for backlink and base LSNs and checkpoint IDs
- found and fixed a bug in `__rec_split_write`, which did not always initialize base checkpoint IDs and LSNs correctly (and left them at 0)
- updated `wt_palm_decode`

As this is a bigger change, I'd appreciate if you could please review the new changes.

quchenhao commented on src/reconcile/rec_write.c:
```diff
@@ -2445,7 +2445,15 @@ __rec_split_write(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_REC_CHUNK *chunk
+             * The first delta needs to explicitly initialize the base checkpoint ID and base LSN to
+             * that of the full image. Future deltas then reuse them.
+             */
+            if (multi->block_meta.delta_count == 0) {
```
This feels not right to me. If we want to build a delta, we must have built the full image before. In that code, we should have set the base lsn and base checkpoint id. In line 2477.

pmacko86 replied:
I believe the 0 base checkpoint IDs and LSNs were coming from line 2486:
```
__wt_page_block_meta_assign(session, &multi->block_meta);
```

quchenhao replied:
Inside that, we also init to

```
/* For a new page, everything starts with the current checkpoint id. */
    meta->checkpoint_id = checkpoint_id;
    meta->backlink_checkpoint_id = checkpoint_id;
    meta->base_checkpoint_id = checkpoint_id;
    meta->disagg_lsn = 0;
```

quchenhao replied:
I do agree we need to init here for lsn as we only get lsn after the page is written.

pmacko86 replied:
You were correct: I found a bug in PALM's `plh_get`, which didn't correctly populate page metadata for the first page. Fixed in [499788a](https://github.com/wiredtiger/wiredtiger/pull/11416/commits/499788aa87cd0abf7001e2512207406dea02d03e).

pmacko86 commented:
I found a bug where the backlink/base validation in PALM fails in `test_layered06`, but very rarely, only maybe like 1 in 20 times or so. I disabled the validation and filed SLS-950 to look into it.

If you think that it's better to have the validation enabled, LMK and I will uncomment it before merging.

quchenhao reviewed (approved):
lgtm

ddanderson commented on ext/page_log/palm/palm.c:
```diff
@@ -694,6 +698,39 @@ palm_handle_get(WT_PAGE_LOG_HANDLE *plh, WT_SESSION *session, uint64_t page_id,
+                PALM_KV_ERR(palm, session, EINVAL);
+        }
+#endif
+        (void)last_lsn;
+        (void)last_checkpoint_id;
```
Nit: are these void statements needed?

pmacko86 replied:
Yes, because the compile otherwise complains that the variables have been unused while the diagnostics code is commented out (as they were assigned values, but the values were then never used).

ddanderson replied:
Ah, okay.  Thanks!

ddanderson reviewed (approved):
LGTM.  One nit, take it if you want.
